Source/Core/Core/HW/EXI/BBA/TAPServer.cpp

```diff
@@ -0,0 +1,305 @@
+      return -1;
+    }
+
+    struct sockaddr_in* sin = reinterpret_cast<struct sockaddr_in*>(&ss);
+    sin->sin_addr.s_addr = htonl(sf::IpAddress(destination.substr(0, colon_offset)).toInteger());
+    sin->sin_family = AF_INET;
+    sin->sin_port = htons(stoul(destination.substr(colon_offset + 1)));
+    ss_size = sizeof(*sin);
+#ifndef _WIN32
+  }
+  else
+  {
+    // UNIX socket
+    struct sockaddr_un* sun = reinterpret_cast<struct sockaddr_un*>(&ss);
+    if (destination.size() + 1 > sizeof(sun->sun_path))
+    {
+      ERROR_LOG_FMT(SP1, "Socket path is too long, unable to init BBA\n");
+      return -1;
+    }
+    sun->sun_family = AF_UNIX;
+    strcpy(sun->sun_path, destination.c_str());
+    ss_size = sizeof(*sun);
+#else
+  }
+  else
+  {
+    ERROR_LOG_FMT(SP1, "UNIX sockets are not supported on Windows\n");
+    return -1;
+#endif
+  }
+
+  int fd = socket(ss.ss_family, SOCK_STREAM, (ss.ss_family == AF_INET) ? IPPROTO_TCP : 0);
+  if (fd == -1)
+  {
+    ERROR_LOG_FMT(SP1, "Couldn't create socket; unable to init BBA\n");
+    return -1;
+  }
+
+#ifdef __APPLE__
+  int opt_no_sigpipe = 1;
+  if (setsockopt(fd, SOL_SOCKET, SO_NOSIGPIPE, &opt_no_sigpipe, sizeof(opt_no_sigpipe)) < 0)
+    INFO_LOG_FMT(SP1, "Failed to set SO_NOSIGPIPE on socket\n");
+#endif
+
+  if (connect(fd, reinterpret_cast<sockaddr*>(&ss), ss_size) == -1)
+  {
+    std::string s = Common::LastStrerrorString();
+    INFO_LOG_FMT(SP1, "Couldn't connect socket ({}), unable to init BBA\n", s.c_str());
+    pi_close(fd);
+    return -1;
+  }
+
+  return fd;
+}
+
+bool CEXIETHERNET::TAPServerNetworkInterface::Activate()
+{
+  if (IsActivated())
+    return true;
+
+  m_fd = ConnectToDestination(m_destination);
+
+  INFO_LOG_FMT(SP1, "BBA initialized.");
+  return RecvInit();
+}
+
+void CEXIETHERNET::TAPServerNetworkInterface::Deactivate()
+{
+  pi_close(m_fd);
+  m_fd = -1;
+
+  m_read_enabled.Clear();
+  m_read_shutdown.Set();
+  if (m_read_thread.joinable())
+    m_read_thread.join();
+}
+
+bool CEXIETHERNET::TAPServerNetworkInterface::IsActivated()
+{
+  return (m_fd >= 0);
+}
+
+bool CEXIETHERNET::TAPServerNetworkInterface::RecvInit()
+{
+  m_read_thread = std::thread(&CEXIETHERNET::TAPServerNetworkInterface::ReadThreadHandler, this);
+  return true;
+}
+
+void CEXIETHERNET::TAPServerNetworkInterface::RecvStart()
+{
+  m_read_enabled.Set();
+}
+
+void CEXIETHERNET::TAPServerNetworkInterface::RecvStop()
+{
+  m_read_enabled.Clear();
+}
+
+bool CEXIETHERNET::TAPServerNetworkInterface::SendFrame(const u8* frame, u32 size)
+{
+  {
+    const std::string s = ArrayToString(frame, size, 0x10);
+    INFO_LOG_FMT(SP1, "SendFrame {}\n{}", size, s);
+  }
+
+  // On Windows, the data pointer is of type const char*; on other systems it is
+  // of type const void*. This is the reason for the reinterpret_cast here and
+  // in the other send/recv calls in this file.
+  u8 size_bytes[2] = {static_cast<u8>(size), static_cast<u8>(size >> 8)};
```

Comment:
It can, but then we would implicitly be relying on Dolphin being built on little-endian architectures. It seems we already do this in a few places, so maybe it's OK here as well, but I'd prefer not to proliferate that assumption.